lib/Dialect/AIE/IR/AIEDialect.cpp

```diff
@@ -1000,20 +1000,42 @@ LogicalResult xilinx::AIE::MemTileDMAOp::verify() {
 // DMABDOp
 LogicalResult xilinx::AIE::DMABDOp::verify() {
   if(getDimensions()) {
+    ::mlir::MemRefType buffer = getBuffer().getType();
+    if(!buffer.getElementType().isInteger(32)) {
+      // The AIE2 specification prescribes that multi-dimensional address
+      // generation creates addresses to 32 bit words. Hence, stepSize and wrap
+      // refer to 32 bit words. To avoid confusion, we disallow using multi-
+      // dimensional BDs with other memrefs.
+      return emitOpError() << "Multi-dimensional buffer descriptors are only "
+                              "supported for 32 bit integer elements.";
+    }
+    uint64_t base_addr = getOffset();
+    uint64_t memref_size = 1; 
```

Comment:
Yes let me add some FileCheck tests

Reply:
Phil and me also discussed making stepsize and wrap byte-sized arguments, and throwing an error if the user passes in values that are not divisible by four. That may be cleaner.

Reply:
Actually, I think this is cleaner and more consistent with the other arguments of the dmaBD, which are also referenced to the size of the memref element type. From a generality perspective, it would be nice to remote the limitation on the element type here, but this is probably good enough for now.

Reply:
I can see arguments for doing it both ways.  As to my understanding, the limitation that the wrap and stepsize is in 4 byte increments is architectural. I think I will implement the following:

The stepsize and wrap are in increments of the element type. However, if the element type is smaller than 4 bytes, I will also verify that the stepsize/wrap are multiples of 4 bytes.

For example: i8 with step size of 2 --> error,
i8 with step size of 4 --> ok
i16 with step size of 2 --> ok
i16 with step size of 1 --> error
i32 with any step size --> ok

Does that seem reasonable?

Reply:
That makes alot of sense to me.

Reply:
Well, except for the "but this won't be the case".  Either they are supported and work as expected, or they don't.  There's some combination of strides which would capture 4 consecutive bytes, followed by 4 consecutive bytes somewhere else, etc.  That's something which should, presumably, be legal.  but if I ask for i8 with a step size of 4 and I don't actually *get* that, then I don't think it's reasonable.

Reply:
The way I understand the hardware (and please correct me if I'm wrong), the DMA _always_ transfers 32 bit words at a time. I'll double-check this.

If it is true, the more I think of this, the harder I think it is to use the multi-dimensional addressing features meaningfully for smaller word sizes. If the hardware transfers 4 bytes at a time and I'm trying to work on smaller units, it's probably going to overwrite some stuff I wouldn't want it to.

For example, consider an array `%buf` that is initialized to all-zero, and I'm reading from a stream that is all `0xFF`s with the following buffer descriptor:

```
AIE.dmaBd(<%buf : memref<8xi8>, 0, 8>, 0, [<4, 8>])
```

I might expect something like this (every fourth byte is set according to stepsize):
```
0xFF,  0,  0,  0,  0xFF,  0,  0,  0
```
but (assuming above is correct) I will get
```
0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF
```

So maybe, then, I should issue an error if multi-dim features are used on <i32 arrays. For >i32 arrays, we could have the dimensions be in units of elemental size. So for an i32 array, one unit of stride is 4 bytes, for an i64 arrays, one unit of stride is 8 bytes.

Reply:
I'd like to be able to use the multi-dimensional BDs with non-i32 element types and have the compiler assume I know what I'm doing (i.e. not disallow it, sanity error checking is OK). In my opinion the wrap and step should expose the hardware registers directly without trying to do any conversion (e.g. no bytes -> words conversion). This is the cleanest and doesn't require casting to use multi-dimensional BDs in the general case (non-i32 data). The feature is much much less useful if only i32 memrefs are supported.

Reply:
Note that at the hardware level (what you pass into the registers), wrap and step size are i32/4 byte increments. You cannot do wrap/step size at byte granularity.

Reply:
I'd remove the warning too. IMHO non-i32 data is the common case and shouldn't issue a warning.

Reply:
So let's say a user has two i8 arrays

```
dst = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

and

```
src = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
```

and they sets up BDs, with the goal of copying some elements from `src` into `dst`.

Let's say they set the stride of the BD to 4.

They may expect

```
dst = [1, 0, 0, 0, 5, 0, 0, 0, 9, 0, 0, 0, 13, 0, 0, 0]
```

i.e. copy every fourth element, but they will get

```
dst = src
```

since four byte words are copied each time.

With data type  `i8` and stride `4`, isn't that highly unintuitive behavior?

Reply:
> isn't that highly unintuitive behavior?

It doesn't matter if it's intuitive. That's what the hardware does and at this level of aie dialect we should expose the hardware to the user.